### PR TITLE
[SPARK-8240][SQL] string function: concat

### PR DESCRIPTION
**Pull request by rxin**

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 1]
```diff
@@ -208,169 +208,6 @@ class DataFrameFunctionsSuite extends QueryTest {
       Row(2743272264L, 2180413220L))
   }
 
-  test("Levenshtein distance") {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: these were moved into StringFunctionsSuite.scala.

**Comment by rxin** [comment]
cc @davies

File: sql/core/src/test/scala/org/apache/spark/sql/StringFunctionsSuite.scala [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types.Decimal
+
+
+class StringFunctionsSuite extends QueryTest {
+
+  private lazy val ctx = org.apache.spark.sql.test.TestSQLContext
+  import ctx.implicits._
+
+  test("string concat") {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/StringFunctionsSuite.scala]: this is the only new test case

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 3]
```diff
@@ -21,7 +21,9 @@
 import java.io.Serializable;
 import java.io.UnsupportedEncodingException;
 
+import org.apache.spark.unsafe.PlatformDependent;
 import org.apache.spark.unsafe.array.ByteArrayMethods;
+import scala.compat.Platform;
```
Comment (rxin) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: note to self: remove this

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala [hunk 4]
```diff
@@ -27,6 +27,40 @@ import org.apache.spark.sql.catalyst.expressions.codegen._
...
+ * Input expressions that are evaluated to nulls are skipped.
+ *
+ * For example, `concat("a", null, "b")` is evaluated to `"ab"`.
+ */
+case class Concat(children: Seq[Expression]) extends Expression with ImplicitCastInputTypes {
+
+  override def inputTypes: Seq[AbstractDataType] = Seq.fill(children.size)(StringType)
+  override def dataType: DataType = StringType
+
+  override def nullable: Boolean = children.exists(_.nullable)
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: We never return null, so this should be `false`

**Comment by rxin** [comment]
OK I'm going to merge this since it blocks my other PR. Note that our concat semantics is different from Hive's w.r.t. null inputs. We can revisit and change that before the release too.

**Pull request closed without merging**